components/match2/wikis/hearthstone/match_group_util_custom.lua

```diff
@@ -0,0 +1,167 @@
+---
+-- @Liquipedia
+-- wiki=hearthstone
+-- page=Module:MatchGroup/Util/Custom
+--
+-- Please see https://github.com/Liquipedia/Lua-Modules to contribute
+--
+
+local Array = require('Module:Array')
+local Logic = require('Module:Logic')
+local Lua = require('Module:Lua')
+local String = require('Module:StringUtils')
+local Table = require('Module:Table')
+
+local MatchGroupUtil = Lua.import('Module:MatchGroup/Util')
+-- can not use `Module:OpponentLibraries`/`Module:Opponent/Custom` to avoid loop
+local Opponent = Lua.import('Module:Opponent')
+
+local CustomMatchGroupUtil = Table.deepCopy(MatchGroupUtil)
+
+---@class HearthstoneMatchGroupUtilSubmatch
+---@field games MatchGroupUtilGame[]
+---@field opponents GameOpponent[]
+---@field resultType ResultType
+---@field scores table<number, number>
+---@field subgroup number
+---@field walkover WalkoverType
+---@field winner number?
+---@field header string?
+
+---@class HearthstoneMatchGroupUtilMatch: MatchGroupUtilMatch
+---@field submatches StormgateMatchGroupUtilSubmatch[]?
+---@field isTeamMatch boolean
+
+---@param record table
+---@return HearthstoneMatchGroupUtilMatch
+function CustomMatchGroupUtil.matchFromRecord(record)
+	local match = MatchGroupUtil.matchFromRecord(record) --[[@as HearthstoneMatchGroupUtilMatch]]
+
+	-- Adjust game.opponents by looking up game.opponents.players in match.opponents
+	Array.forEach(match.games, function(game)
+		game.opponents = CustomMatchGroupUtil.computeGameOpponents(game, match.opponents)
+	end)
+
+	match.isTeamMatch = Array.any(match.opponents, function(opponent)
+		return opponent.type == Opponent.team end
+	)
+
+	if match.isTeamMatch then
+		-- Compute submatches
+		match.submatches = Array.map(
+			CustomMatchGroupUtil.groupBySubmatch(match.games),
+			function(games) return CustomMatchGroupUtil.constructSubmatch(games) end
+		)
+
+		local extradata = match.extradata
+		---@cast extradata table
+		Array.forEach(match.submatches, function (submatch)
+			submatch.header = Table.extract(extradata, 'subgroup' .. submatch.subgroup .. 'header')
+		end)
+	end
+
+	return match
+end
+
+---@param game MatchGroupUtilGame
+---@param matchOpponents standardOpponent[]
+---@return table[]
+function CustomMatchGroupUtil.computeGameOpponents(game, matchOpponents)
+	return Array.map(game.opponents, function (opponent, opponentIndex)
+		return Table.merge(opponent, {
+			players = Array.map(game.opponents[opponentIndex].players or {},function (player, playerIndex)
+				if Logic.isEmpty(player) then return nil end
```

Comment:
not needed i think
```suggestion
```

Reply:
This keeps the players array without empty elements